src/rgw/rgw_cache.cc

```diff
@@ -344,3 +359,424 @@ ObjectCache::~ObjectCache()
+}
+
+int DataCache::io_write(bufferlist& bl ,unsigned int len, std::string oid) {
+
+  ChunkDataInfo*  chunk_info = new ChunkDataInfo;
+
+  std::string location = cct->_conf->rgw_datacache_persistent_path + oid; /* replace tmp with the correct path from config file*/
+  FILE *cache_file = 0;
+  int r = 0;
+
+  cache_file = fopen(location.c_str(),"w+");
+  if (cache_file <= 0)
+  {
+    ldout(cct, 0) << "ERROR: DataCache::open file has return error " << r << dendl;
+    return -1;
+  }
+
+  r = fwrite(bl.c_str(), 1, len, cache_file);
+  if (r < 0) {
+    ldout(cct, 0) << "ERROR: DataCache::write: write in file has return error " << r << dendl;
+  }
+
+  fclose(cache_file);
+
+  /*update cahce_map entries for new chunk in cache*/
+  cache_lock.Lock();
+  chunk_info->oid = oid;
+  chunk_info->set_ctx(cct);
+  chunk_info->size = len;
+  cache_map.insert(pair<string, ChunkDataInfo*>(oid, chunk_info));
+  cache_lock.Unlock();
+
+  return r;
+}
+
+void _cache_aio_write_completion_cb(sigval_t sigval) {
+
+  cacheAioWriteRequest *c = (cacheAioWriteRequest *)sigval.sival_ptr;
+  c->priv_data->cache_aio_write_completion_cb(c);
+}
+
+
+void DataCache::cache_aio_write_completion_cb(cacheAioWriteRequest *c){
+
+  ChunkDataInfo  *chunk_info = NULL;
+
+  ldout(cct, 0) << "engage: cache_aio_write_completion_cb oid:" << c->oid <<dendl;
+
+  /*update cahce_map entries for new chunk in cache*/
+  cache_lock.Lock();
+  outstanding_write_list.remove(c->oid);
+  chunk_info = new ChunkDataInfo;
+  chunk_info->oid = c->oid;
+  chunk_info->set_ctx(cct);
+  chunk_info->size = c->cb->aio_nbytes;
+  cache_map.insert(pair<string, ChunkDataInfo*>(c->oid, chunk_info));
+  cache_lock.Unlock();
+
+  /*update free size*/
+  eviction_lock.Lock();
+  free_data_cache_size -= c->cb->aio_nbytes;
+  outstanding_write_size -=  c->cb->aio_nbytes;
+  lru_insert_head(chunk_info);
+  eviction_lock.Unlock();
+  c->release();
+}
+
+int DataCache::create_aio_write_request(bufferlist& bl, unsigned int len, std::string oid){
+
+  struct cacheAioWriteRequest *wr= new struct cacheAioWriteRequest(cct);
+  int r=0;
+  if (wr->create_io(bl, len, oid) < 0) {
+    ldout(cct, 0) << "engage: Error create_aio_write_request" << dendl;
+    goto done;
+  }
+  wr->cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+  wr->cb->aio_sigevent.sigev_notify_function = _cache_aio_write_completion_cb;
+  wr->cb->aio_sigevent.sigev_notify_attributes = NULL;
+  wr->cb->aio_sigevent.sigev_value.sival_ptr = (void*)wr;
+  wr->oid = oid;
+  wr->priv_data = this;
+
+  if((r= ::aio_write(wr->cb)) != 0) {
+    ldout(cct, 0) << "ERROR: aio_write "<< r << dendl;
+    goto error;
+  }
+  return 0;
+
+error:
+  wr->release();
+done:
+  return r;
+}
+
+void DataCache::put(bufferlist& bl, unsigned int len, std::string oid){
+
+  int r = 0;
+  long long freed_size = 0, _free_data_cache_size = 0, _outstanding_write_size = 0;
+
+  ldout(cct, 20) << "Engage1: We are in DataCache::put() and oid is: " << oid <<dendl;
```

Comment:
what does Engage1 mean?

Reply:
it originates from the original D3N PR, renamed the `Engage*:` to `DataCache:` in all the log refrences

Reply:
all mentions of `Engage1` were removed